javascripts/2048.js

```diff
@@ -1,32 +1,406 @@
+  }
+
+  for (var i = 0; i < this.length; i++) {
+    // Check if we have nested arrays
+    if (this[i] instanceof Array && array[i] instanceof Array) {
+      // recurse into the nested arrays
+      if (!this[i].equals(array[i])) {
+        return false;
+      }
+    } else if (this[i] != array[i]) {
+      // Warning - two different object instances will never be equal: {x:20} != {x:20}
+      return false;
+    }
+  }
+  return true;
+
+}
+
 $(document).ready(function() {
   console.log('ready!');
+  prepareBoard();
+
   $('body').keydown(function(event){
     var arrow_keys = [37, 38, 39, 40];
-    if(arrow_keys.indexOf(event.which) > -1) {
-      var tile = $('.tile');
-      moveTile(tile, event.which);
+    var direction = event.which;
+    if(arrow_keys.indexOf(direction) > -1) {
+      makeTurn(direction);
       event.preventDefault();
     }
   })
+
+  $('.button.new-game').click(function() {
+    location.reload();
+  })
 })
 
-function moveTile(tile, direction) {
-  var new_tile_value = tile.attr("data-val") * 2;
-  tile.attr("data-val", new_tile_value);
-  tile.text(new_tile_value);
-
-  switch(direction) {
-    case 38: //up
-      tile.attr("data-row","r0");
-      break;
-    case 40: //down
-      tile.attr("data-row","r3");
-      break;
-    case 37: //left
-      tile.attr("data-col","c0");
-      break;
-    case 39: //right
-      tile.attr("data-col","c3");
-      break;
+function prepareBoard() {
+  addTile();
+  addTile();
+}
+
+
+function tileSelectorText(row, col) {
+  return ".tile[data-row=\"" + row + "\"][data-col=\"" + col + "\"]";
+}
+
+function collectEmptySpaces() {
+  var emptySpaces = [];
+  var row = null;
+  var col = null;
+  var tileLocation = null;
+  // figure out all empty spots
+  for (var i = 1; i < 5; i++) { // loop through rows
+    row = i;
+    for (var j = 1; j < 5; j++) { // loop through cols
+      col = j;
+      tileLocation = $(tileSelectorText(row, col));
+      if (tileLocation.length == 0) {
+        emptySpaces.push([row, col]);
+      }
+    }
+  }
+  return emptySpaces;
+}
+
+function addTile() {
+  var tile = $("<div data-row='' data-col='' data-val=''></div>");
+  var dataVal = Math.random() < 0.04 ? 4 : 2;
+  var emptySpaces = collectEmptySpaces(); // guard for if no empty spaces?
+
+  // function to pick row and column number
+  var randomLocation = emptySpaces[Math.floor(Math.random() * emptySpaces.length)];
+
+  tile.addClass("tile");
+  tile.attr("data-row", randomLocation[0]);
+  tile.attr("data-col", randomLocation[1]);
+  tile.attr("data-val", dataVal);
+  tile.text(dataVal);
+  tile.addClass("new");
+
+  $("#gameboard").append(tile);
+  // remove new class after animation is complete
+  tile.on("animationend", function() {
+    tile.removeClass("new");
+  })
+}
+
+function makeTurn(direction) {
+  var tiles = $(".tile"); // in order by the html
+  var turnType = rowOrColumn(direction);
+  var turnOppositeType = (turnType == "data-row") ? "data-col" : "data-row";
```

Comment:
Note that `rowOrColumn` may have returned an empty string at this point if for any reason `direction` was not between 37 and 40. In that scenario `turnOppositeType` would arbitrarily receive `data-row`. For safety the following could be added:
```
if (!turnType) {
  return;
}
```